feat(benchmark)!: rewrite the public API (#10113)
vitest-dev/vitest: docs/guide/benchmarking.md

Reviewer: The content looks good, but I was puzzled by how I can actually get started. Probably first section or new section should mention the activation switch. Codex told me something like this:

---

| Command/config | Regular test project | Benchmark project |
|---|---:|---:|
| `vitest` / `vitest run`, no `benchmark.enabled` | runs | not created |
| `vitest` / `vitest run`, with `benchmark.enabled: true` | runs | runs too |
| `vitest bench` | filtered out | runs only |

A benchmark project is a cloned project using `benchmark.include` instead of `test.include`, defaulting to `*.bench.ts` / `*.benchmark.ts`.

So practically:

- `vitest run` runs normal tests.
- `vitest run` + `benchmark.enabled: true` runs normal tests and benchmark files.
- `vitest bench` runs benchmark files only, no config needed.

And `{ bench }` itself does not decide collection. It only works once the file is already in the benchmark project.

    bench('lib2', () => { lib2() }),
  )

  expect(result.get('lib1')).toBeFasterThan(result.get('lib2'))
})
```

The `delta` option specifies the minimum relative difference required for the assertion to pass. This helps avoid flaky tests caused by benchmark noise:

```ts
// lib1 must be at least 10% faster than lib2
expect(result.get('lib1')).toBeFasterThan(result.get('lib2'), {
  delta: 0.1,
})

// lib2 must be at least 20% slower than lib1
expect(result.get('lib2')).toBeSlowerThan(result.get('lib1'), {
  delta: 0.2,
})
```

You can also assert absolute performance using standard matchers:

```ts
test('parsing is fast enough', async ({ bench }) => {
  const result = await bench('parse', () => {
    parse(largeInput)
  }).run()

  expect(result.throughput.mean).toBeGreaterThan(10_000)
})
```

## Retries

Since benchmarks can be noisy, use the `retry` option to automatically retry failing benchmark tests:

```ts
test('performance comparison', { retry: 3 }, async ({ bench }) => {
  const result = await bench.compare(
    bench('lib1', () => { lib1() }),
    bench('lib2', () => { lib2() }),
  )

  expect(result.get('lib1')).toBeFasterThan(result.get('lib2'))
})
```

> ## Baselines

Reviewer: I'm not sure new baseline API is convenient for common workflow. 

For example, previous approach with throw-away output json + compare allowed switching around branch to get the comparison like https://main.vitest.dev/config/benchmark.html#benchmark-outputJson

```sh
# save main branch's result
git checkout main
vitest bench --outputJson main.json

# change a branch and compare against main
git checkout feature
vitest bench --compare main.json
```

How does the new API (with `__benchmark__` encouraged to be committed) allow similar workflow?

Author: You can define both:

```ts
await bench.compare(
  bench('always run', () => { code() })
  bench.withBaseline('stored bench', () => { code() })
)
```

Instead of doing it in CLI level, you define it in test

Use `bench.withBaseline()` to store benchmark results on disk and compare against them in future runs, similar to how [snapshot testing](/guide/snapshot) works:

```ts
test('no performance regression', async ({ bench }) => {
  const result = await bench.withBaseline('parse', () => {
    parse(largeInput)
  }).run()
})
```

- **First run**: Executes the benchmark and stores the result to a `__benchmarks__/` directory next to the test file.
- **Subsequent runs**: Executes the benchmark and compares against the stored result, reporting any regressions.

> - **Updating baselines**: Use the `--update-baselines` flag to overwrite stored baselines with fresh results.

Reviewer: I got confused / mislead by `Subsequent runs` claim. It reads like `bench.withBaseline + run` alone somehow surfaces the baseline comparison, but it's meant to combined with general `bench.compare` pattern to show comparison table in reporter.

I know the API is servicing as designed but the specific baseline workflow probably requires more comprehensive explainer like maybe with step by step example.

---

Also another perspective is that, how about providing ergonomic wrapper API to combine `compare + withBaseline (+ toBeFasterThan)`?

Author: I'm not sure about adding more and more wrappers, I'd rather we design an API that's easy to combine together. I think `withBaseline` is flexible enough, but I am willing to see the exact proposal you are thinking of (how the code looks like?)

Reviewer: I think the primitive is right, so not a blocker. The documentation wise though, it should imply this primitive can be used together with `compare` API or manual run result comparison.

---

For API wise (or higher level integration suggestion from doc), I spend sometime but struggling how this is meant to be used, so it was just an open question from me.

Exploring some concrete use case to target a tangible goal and here is one story. I was thinking about how to setup CI workflow to compare bench result from main against each PR. One integration example I was thinking would be  something like:

- 1. have ci job like this

```yaml
jobs:
  # either generate fresh main checkout as baseline in each pr job
  # or assume main branch job generated such artifact
  benchmark-baseline:
    steps:
      - uses: actions/checkout@v4
        with:
          ref: ${{ github.base_ref || 'main' }}
      - run: pnpm vitest bench --update-baselines
      - uses: actions/upload-artifact@v4
        with:
          name: benchmark-baselines
          path: '**/__benchmarks__/*.json'  # side note. this could be simpler if we employ .vitest/benchmark/*

  # run pr checkout bench against main baseline artifact
  benchmark-pr:
    needs: benchmark-baseline
    steps:
      ...
      - uses: actions/download-artifact@v4
        with:
          name: benchmark-baselines
      - run: pnpm vitest bench
```

- 2. the latter bench run wants to surface obvious large regression (assuming github actions environment aren't that stable for precise benchmarking, but maybe good enough to catch bad regression)

Some characteristics for this specific use case are:
- benchmark artifact is a throw-away artifact, so no need to be commited. It could be even simplified if it's consolidated to root `.vitest/benchmark/*` instead of `__benchmark__`
- this use case only requires the simplest combination of `bench.withBaseline + bench.compare` such as:

```js
const result = await bench.compare(
  bench('current', () => current()),
  bench.withBaseline('baseline', () => current()),
)
```

- The pattern of how to present compare `result` can be also somewhat standardized. Being able to define threshold inside code itself (i.e. per bench case) instead of global config is likely still important. But, it feels like it should end up as reporter-accessible metadata so that reporter can eventually surface it specifically for the common pattern.

So given that assumption, I could see API like this seems more ergonomic to fit the story:

```js
// conceptually a sugar of compare + withBaseline + toBeFasterThan
// but additionally extra metadata (e.g. bench name, threshold) are available through
// the assertion error property or maybe as annotation so builtin (or 3rd party) reporter can associate
// with raw benchmark result.
await bench.compareBaseline("my-code", { threshold: etc... }, () => someFn())
```

Author: Proposal:

- step by step example of generating baselines
- workflow that commits updated baselines (when/before pr is merged?)

```ts
await bench('fn', () => current()).run({ writeResult: './stored-artifact.json' })

await bench.compare(
  // bench('fn', { writeResult: './stored-artifact.json' }, () => current()),
  bench.from('./stored-artifact.json'),
  bench.from(() => fetch('./stored-artifact.json').json()),
)
```

Reviewer: The minimal example I have in mind for baseline use case demo:

```js
test("demo", ({ bench }) => {
  if (import.meta.env.WRITE_BENCH) {
    const baseline = bench("baseline", { writeResults: "./my-bench.json" }, () => fn())
    await baseline.run()
  }  else {
    const baseline = bench.from("baseline", "./my-bench.json")
    await bench.compare(bench(name, () => fn()), baseline)
  }
})
```

Or extracting it as simple wrapper utility:

```js
test("add", ({ bench }) => {
  await benchCompareBaseline(bench, "add", () => 2 + 2, "./add.json")
})

test("mul", ({ bench }) => {
  await benchCompareBaseline(bench, "mul", () => 2 * 2, "./mul.json")
})

function benchCompareBaseline(bench, name, fn, filepath) {
  if (import.meta.env.WRITE_BENCH) {
    const baseline = bench("baseline", { writeResults: filepath }, () => fn())
    await baseline.run()
  }  else {
    const baseline = bench.from("baseline", { filepath })
    await bench.compare(bench(name, () => fn()), baseline)
  }
}
```

> 
> Baselines work inside `bench.compare()` too. You can mix regular and baseline benchmarks:
> 
> ```ts
> test('compare against baseline', async ({ bench }) => {
>   const result = await bench.compare(
>     bench('current implementation', () => { current() }),
>     bench.withBaseline('previous implementation', () => { previous() }),
>   )

Reviewer: Is this intended to be both `current`?

```suggestion
  const result = await bench.compare(
    bench('current implementation', () => { current() }),
    bench.withBaseline('previous implementation', () => { current() }),
  )
```

To further improve stability:

- Use the [`retry`](#retries) option to automatically rerun flaky benchmark assertions.
- Use the [`delta`](#asserting-performance) option in `toBeFasterThan` / `toBeSlowerThan` to allow for acceptable variance.
- Avoid running benchmarks alongside CPU-intensive processes.
- Close browsers, IDEs, and other applications that compete for CPU time.

### Dead Code Elimination

JavaScript engines can optimize away code that has no observable side effects. If your benchmark function doesn't use its result, the engine may skip the computation entirely, producing misleadingly fast numbers:

```ts
test('parsing', async ({ bench }) => {
  // BAD: the engine may eliminate the work
  await bench('parse', () => {
    JSON.parse(input)
  }).run()

  // GOOD: the result is consumed
  await bench('parse', () => {
    const result = JSON.parse(input)
    doSomething(result)
  }).run()
})
```

This applies to all engines (V8, JavaScriptCore, SpiderMonkey) but is especially aggressive in V8's TurboFan and JavaScriptCore's FTL compiler tiers.

> ### Module Runner Overhead

Reviewer: We can also suggest pre-built library as external when comparing own library against others. Also for pre-built one, users can disable module runner potentially. Related https://github.com/vitest-dev/vitest/issues/6543

---
title: Benchmarking | Guide
---

# Benchmarking

Vitest lets you write benchmarks alongside your tests using the `bench` fixture from the [test context](/guide/test-context). It is not a top-level import from `vitest`. Benchmarks are powered by [Tinybench](https://github.com/tinylibs/tinybench) and are defined inside regular `test()` calls, giving you access to the full power of Vitest's test runner: retries, lifecycle hooks, filtering, and assertions.

## Defining a Benchmark

Use the `bench` fixture to define a benchmark. Call `.run()` to execute it:

```ts
import { expect, test } from 'vitest'

test('parsing performance', async ({ bench }) => {

>   const result = await bench('parse', () => {
>     JSON.parse('{"key":"value"}')
>   }).run()

Reviewer 2: Below it states following:

> The `bench()` function registers a benchmark without executing it. Calling `.run()` runs the benchmark and returns the result.

To make that even more clear in this example, I would suggest following:

```suggestion
  const jsonParseBenchmark = bench('parse', () => {
    JSON.parse('{"key":"value"}')
  })
  const result = await jsonParseBenchmark.run()
```

Author: I prefer the current example because that is how it's going to be used and spiting it into two variables is not really useful

### `vitest` (default)

Without [`benchmark.enabled`](/config/#benchmark-enabled), the `vitest` command only runs regular tests. Benchmark files are ignored entirely. This is the default and the right choice for day-to-day development, since benchmarks are slow and noisy and shouldn't run on every save.

### `vitest` with `benchmark.enabled`

Set `benchmark.enabled: true` in your config to run benchmarks together with regular tests:

```ts [vitest.config.ts]
import { defineConfig } from 'vitest/config'

export default defineConfig({
  test: {
    benchmark: {
      enabled: true,
    },
  },
})
```

With this config, `vitest` runs your regular tests first, then runs the benchmarks in a separate isolated group (so benchmark execution never overlaps with test execution and adds noise to results). Useful in CI when you want a single command to validate correctness and performance.

### `vitest bench`

> The `bench` subcommand runs benchmarks only and skips regular tests:

Reviewer 2: Nit:

```suggestion
The `bench` subcommand runs only benchmarks and skips regular tests:
```

```bash
vitest bench
```

This implicitly enables `benchmark.enabled` for the run, so you don't need to set it in the config. Like the `vitest` command, it accepts filename filters and `-t`/`--testNamePattern` to narrow the run:

```bash
# only benchmarks in files matching "parser"
vitest bench parser

# only benchmarks whose test name matches "JSON"
vitest bench -t JSON
```

> ## Comparing Benchmarks

Reviewer 2: Haven't yet read through the whole PR but: Is it possible to save benchmark results with specific name, and then compare against those without ever updating them?

```ts
import { test } from 'vitest'
import { customParser } from '../dist';

test('compare against v1', async ({ bench }) => {
  const v1Results = await bench.load("v1-results.json") // or something?

  const result = await bench.compare(
    bench('current v2', () => {
      customParser(input)
    }),
    v1Results,
  )
})
```

Author: What would `load` do if there is nothing to load? `bench.withBaseline` has a fallback to run and generate the result

Reviewer 2: It would be same as `withBaseline`, but it also would make sure that on file system there is a baseline that is never updated, even when `--update-baselines` is used. If result matching the name (`v1-results.json` in example) is not found, it would just error out.

Author: How would you generate it then?

Reviewer 2: Needs a way to write benchmarks to file system. 🤔

Author: The API in this PR is clear - you use results for comparison with `expect`, and you use `bench` to show the comparison table (`.compare`, `.run`), I don't understand what `.load` adds that `withBaseline` doesn't cover. What does it return? How do you generate that file if it only reads it? If another function generates it, why should this one return it? _When_ is it generated? How do you use it? Do you show the table? 

`.compare` defines the order and other options of benchmarks (`tinybench.add().add().add().run(options)`) - we don't do `await bench, await bench, await bench` for that exact reason

`.load` just doesn't fit in anywhere

Reviewer 2: Is there a way for me to use this API and see how my project's performance has changed over time, in multiple versions? So if I was using this feature in my project over three major versions, could I somehow see how my project's performance was in v1, v2 and currently for example?

Or is it just possible to see how it's performing right now, and in previous baseline?

Author: Sure it’s possible, withBaseline reads the file from the disk. If it was generated in v1, it will show that. How do you think withBaseline works? Are docs confusing?

Reviewer 2: Can you write an example of test case that would compare `import { customParser } from '../dist';` function against its performance in `v1`, `v2` and current? How does `--update-baselines` ignore updating those?

Author: Why would you run update-baselines if you don’t want to update it? Just don’t update baselines

You run it once, it’s generated and it stays there. It would look like 3 bench.withBaseline calls inside bench.compare that were added there gradually 

If you want to be strict, just throw an error in the baseline function

Author: ```ts
await bench.compare(
  bench.withBaseline('v1', impl)
)
// run, generates v1
await bench.compare(
  bench.withBaseline('v1',() => { throw new Error() })
  bench.withBaseline('v2', impl)
)
// run, generates v2
```

```ts
test('benchmarks with setup', async ({ bench }) => {
  const result = await bench.compare(
    bench('with-cache', () => {
      readFromCache()
    }),
    bench(
      'without-cache',
      { beforeEach: () => clearCache() },
      () => { readFromDisk() },
    ),
  )
})
```

## Comparing Across Projects

When your workspace defines multiple projects (e.g., different browsers or runtimes), you can use `bench.perProject()` to compare how the same benchmark performs across all of them. Instead of printing results inline per project, Vitest collects them and prints a single comparison table at the end of the test run.

```ts
import { test } from 'vitest'

test('simple example', async ({ bench }) => {
  await bench.perProject('1 + 1', () => {
    1 + 1
  }).run()
})
```

The same test file runs in each project (chromium, firefox, webkit, etc.), and Vitest groups the results:

<<< ./snippets/benchmark-per-project.ansi

`bench.perProject()` returns a `BenchRegistration` just like `bench()`, so you call `.run()` to execute it. You can also mix it with `bench.compare()`:

```ts

> test('compare implementations across browsers', async ({ bench }) => {
>   await bench.compare(
>     bench.perProject('JSON.parse', () => {
>       JSON.parse('{"key":"value"}')
>     }),
>     bench('custom parser', () => {
>       customParse('{"key":"value"}')
>     }),
>   )
> })

Reviewer 2: Did you consider having `perProject` as configuration property instead? I find it a bit weird as property in `bench` 🤔 

```suggestion
test('compare implementations across browsers', async ({ bench }) => {
  await bench.compare(
    bench('JSON.parse', { perProject: true }, () => {
      JSON.parse('{"key":"value"}')
    }),
    bench('custom parser', () => {
      customParse('{"key":"value"}')
    }),
  )
})
```

Author: Yes, I went with an option because bench options are just tinybench options (simpler mental image, in my opinion). I am open to reconsider

  ```ts
  test('sorting', async ({ bench }) => {
    const original = Array.from({ length: 10000 }, () => Math.random())
    let data: number[]

    // BAD: allocates a new array every iteration, GC adds noise
    await bench('sort', () => {
      const data = Array.from({ length: 10000 }, () => Math.random())
      data.sort()
    }).run()

    // GOOD: pre-allocate, copy in beforeEach
    await bench(
      'sort',
      () => { data.sort() },
      {
        beforeEach() {
          data = [...original]
        },
      },
    ).run()
  })
  ```

#### JavaScriptCore (Bun, Safari)

- **Different optimization thresholds**: JSC uses its own JIT tiers (LLInt → Baseline → DFG → FTL) with different inlining and optimization heuristics. A benchmark that is fast on V8 may behave very differently on JSC.
- **Async benchmarks**: Bun's event loop implementation differs from Node.js. If your benchmark involves async operations or timers, results may not be directly comparable across runtimes.

#### Browser

> - **Timer resolution**: Browsers may reduce `performance.now()` precision (e.g., to 100μs or even 1ms) as a Spectre mitigation. This makes very fast operations difficult to measure accurately, so increase iterations to compensate:

Reviewer 2: I'd rather not mention specific vulnerabilities, and keep this generic:

```suggestion
- **Timer resolution**: Browsers may reduce `performance.now()` precision (e.g., to 100μs or even 1ms) as a security mechanism. This makes very fast operations difficult to measure accurately, so increase iterations to compensate:
```

---
title: Benchmarking | Guide
---

# Benchmarking

> Vitest lets you write benchmarks alongside your tests using the `bench` fixture from the [test context](/guide/test-context). It is not a top-level import from `vitest`. Benchmarks are powered by [Tinybench](https://github.com/tinylibs/tinybench) and are defined inside regular `test()` calls, giving you access to the full power of Vitest's test runner: retries, lifecycle hooks, filtering, and assertions.

Reviewer: nit, repeated mention of "not a top-level import from `vitest`" reads a bit off. If this is for migration intent, then I'd either leave it migration.md or the migration intent should be phrased cleary in the small warning block.

## Defining a Benchmark

Use the `bench` fixture to define a benchmark. Call `.run()` to execute it:

```ts
import { expect, test } from 'vitest'

test('parsing performance', async ({ bench }) => {
  const result = await bench('parse', () => {
    JSON.parse('{"key":"value"}')
  }).run()
})
```

> The `bench()` function registers a benchmark without executing it. Calling `.run()` runs the benchmark and returns the result. Vitest will print the benchmark output (ops/sec, mean time, etc.) after the test completes.

Reviewer: I had an impression that bare `bench + run` won't print anything as there wasn't reporter output example other than `compare` and `perProject`. Can we add the bare example output here?

Author: I feel like there are already enough examples, it will be printed the same way as `compare`, just with a single element with it - I can add this as a note to compare result

```ts
test('compare parser versions', async ({ bench }) => {
  const input = '{"key":"value"}'

  await bench.compare(
    bench.from('v1', './benchmarks/parse.v1.json'),
    bench.from('v2', './benchmarks/parse.v2.json'),
    bench(
      'current',
      { writeResult: './benchmarks/parse.current.json' },
      () => customParser(input),
    ),
  )
})
```

To produce a new historical artifact, point a fresh `bench()` at that version's implementation, set `writeResult` to a versioned path (`./benchmarks/parse.v3.json`), run it once, then replace the call with `bench.from('v3', './benchmarks/parse.v3.json')`.

To regenerate the baseline on demand, gate the write behind an environment variable so the same test either refreshes the artifact or compares against it:

```ts
test('compare parser versions', async ({ bench }) => {
  if (import.meta.env.VITE_WRITE_BENCH) {
    const baseline = bench('baseline', { writeResult: './my-bench.json' }, () => fn())
    await baseline.run()
  }
  else {
    const baseline = bench.from('baseline', './my-bench.json')
    await bench.compare(bench('current', () => fn()), baseline)
  }
})
```

> Run `WRITE_BENCH=1 vitest bench` to refresh the stored result, and `vitest bench` to compare the current implementation against it.

Reviewer: `VITE_WRITE_BENCH` and `WRITE_BENCH` are inconsistent. Probably Browser mode only picks up `VITE_...` prefix, so it should be `VITEST_WRITE_BENCH=1 vitest bench`.

### Dead Code Elimination

JavaScript engines can optimize away code that has no observable side effects. If your benchmark function doesn't use its result, the engine may skip the computation entirely, producing misleadingly fast numbers:

```ts
test('parsing', async ({ bench }) => {
  // BAD: the engine may eliminate the work
  await bench('parse', () => {
    JSON.parse(input)
  }).run()

  // GOOD: the result is consumed
  await bench('parse', () => {
    const result = JSON.parse(input)
    doSomething(result)
  }).run()
})
```

This applies to all engines (V8, JavaScriptCore, SpiderMonkey) but is especially aggressive in V8's TurboFan and JavaScriptCore's FTL compiler tiers.

### Module Runner Overhead

By default, Vitest runs tests in Node.js using Vite's module runner (configured by [`experimental.viteModuleRunner`](/config/experimental#experimental-vitemodulerunner)). This transforms all module exports into getters, so every access to an imported binding goes through something like `__vite_ssr_module__.value`. In regular tests this overhead is negligible, but in benchmarks where a function is called millions of times, the getter call itself can dominate the measurement.

> Vitest will print a warning if it detects excessive getter calls, but you should be aware of this when benchmarking imported functions:

Reviewer: There's no mention of `suppressExportGetterWarnings` in the doc. Probably worth mentioning here and `docs/config/benchmark.md`.
